examples/src/high_level_bottle.rs

```diff
@@ -34,7 +34,7 @@ pub fn shape() -> Shape {
     let neck_radius = thickness / 4.0;
     let neck_height = height / 10.0;
 
-    let mut neck = AdHocShape::make_cylinder(dvec3(0.0, 0.0, height), neck_radius, neck_height);
+    let neck = AdHocShape::make_cylinder(dvec3(0.0, 0.0, height), neck_radius, neck_height);
     neck.union(&body);
```

Comment:
This change is now incorrect, as `union()` is no longer a mutable function, but something that returns a new `BooleanShape`.

`neck.union(&body);` now needs to be `let neck = neck.union(&body);`

My suggestion would be to add a `#[must_use]` attribute to all functions which return `BooleanShape` so it's harder to make this mistake.

Reply:
Perhaps we'd also want to add in-place, mutable versions of the boolean operations (`union_mut()` or something like that).

Reply:
Good point, I think this is an artifact of a previous PR.

I actually prefer returning a new object instead of the mutable option as it allows for chaining operations. However, this doesn't cover cases like iterating over a some shapes and union it with another one (though this can be done with fold(), but is not as straight forward). What do you think about something like a `union_shapes()` function which takes an iterator of objects implementing `Into<Shape>`?

On another note, some operations like `fillet()` mutate the object while others like `union()` return a new one. It may be a good idea to select one of the styles to support to make the coding experience more uniform:

Example when using exclusively the first style:
```rust
let outer_box = case_outer_box()
    .subtract(&inner_box)
    .fillet(1.3)
    .union(&top_shelf)
    .union(&bottom_shelf)
    .subtract(&usb_cutout)
    .union_shapes(&SUPPORT_POSTS)
```

Example when using exclusively the second one:
```rust
let mut outer_box = case_outer_box();

outer_box.subtract(&inner_box);
outer_box.fillet(1.3);

outer_box.union(&top_shelf);
outer_box.union(&bottom_shelf);
outer_box.subtract(&usb_cutout);

for support_post in SUPPORT_POSTS {
    outer_box.union(&support_post.shape());
}
```

Which one do you prefer or should we rather keep the mixed approach?

Reply:
I think the first example is nicer, though it would be nice to also support the second style if we can come up with an API that isn't ugly. My only first thought is something like `subtract_mut` and `fillet_mut` but that's still pretty ugly.

Probably for now it's better to support a chainable API, and use the `#[must_use]` attribute so users don't make the mistake of ignoring the result.

Reply:
This was quite something, but I now made almost the entire API chainable and added `#[must_use]` whenever `&self` is used and a new shape object is returned. I also adjusted the examples accordingly and IMO their code became a lot nicer in a lot of places!

Something like `union_shapes()` and the `*_mut` operations should probably best be done in another PR.

Reply:
Very nice! I'll try to get a more thorough review on this soon.

> should probably best be done in another PR

Agreed, this one has blown up a fair amount from my suggestions 😅